fedbiomed/node/requests/_n2n_controller.py

```diff
@@ -126,7 +134,49 @@ async def _HandlerDefault(self, overlay_msg: dict, inner_msg: InnerMessage) -> N
             f"{ErrorNumbers.FB324}: Failed processing overlay message, unknown inner command "
             f"{inner_msg.__class__.__name__}. Do nothing.")
 
-    async def _HandlerKeyRequest(self, overlay_msg: dict, inner_msg: InnerMessage) -> dict:
+    async def _HandlerChannelRequest(self, overlay_msg: OverlayMessage, inner_msg: InnerMessage) -> dict:
+        """Handler called for ChannelSetupRequest message.
+
+        Args:
+            overlay_msg: Outer message for node to node communication
+            inner_msg: Unpacked inner message from the outer message
+
+        Returns:
+            A `dict` with overlay reply message
+        """
+        # we assume the data is properly formatted
+        inner_resp = ChannelSetupReply(
+            request_id=inner_msg.request_id,
+            node_id=environ['NODE_ID'],
+            dest_node_id=inner_msg.node_id,
+            public_key=await self._overlay_channel.get_local_public_key(inner_msg.node_id))
+
+        overlay, salt = await self._overlay_channel.format_outgoing_overlay(
+            inner_resp,
+            overlay_msg.researcher_id,
+            True
+        )
+        overlay_resp = OverlayMessage(
+            researcher_id=overlay_msg.researcher_id,
+            node_id=environ['NODE_ID'],
+            dest_node_id=inner_msg.node_id,
+            overlay=overlay,
+            setup=True,
+            salt=salt)
+
+        return { 'overlay_resp': overlay_resp }
+
+
+    async def _FinalChannelReply(self, inner_msg: InnerMessage) -> None:
+        """Final handler called for ChannelSetupReply message.
+
+        Args:
+            inner_msg: received inner message
+        """
+        await self._overlay_channel.set_distant_key(inner_msg.get_param('node_id'), inner_msg.get_param('public_key'))
```

Comment:
It would be nice to use also pending requests here to be coherent to other replies and simplify future refactor.

Reply:
As discussed: we need a task (asyncio) sync mechanism. Using a thread sync mechanism caused a deadlock, which I addressed with the refactoring moving `Overlay` in the `N2nRouter` thread and converting it to use asyncio.